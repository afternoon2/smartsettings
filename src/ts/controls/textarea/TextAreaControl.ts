import { Control } from '../control/Control';
import { InternalState, TextAreaControlOptions, Listener, ListenerUpdate } from '../../types';

import Styles from '../../../sass/textarea.sass';

export type TextAreaControlProps = {
  id: string,
  options: TextAreaControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class TextAreaControl extends Control {
  public controlElement: HTMLTextAreaElement;

  protected static defaultRows: number = 4;
  protected static defaultCols: number = 10;

  protected static template = (state: InternalState): string => `
    <textarea
      wrap="hard"
      class="${Styles.textarea}"
      cols="${state.cols || TextAreaControl.defaultCols}"
      rows="${state.rows || TextAreaControl.defaultRows}"
      ${state.disabled ? 'disabled="true"' : ''}
      ${state.minLength ? `min-length="${state.minLength}"` : ''}
      ${state.maxLength ? `max-length="${state.maxLength}"` : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
      ${state.readonly ? 'readonly="true"' : ''}
      ${state.autocomplete ? 'autocomplete="true"' : ''}
    >${state.value || ''}</textarea>
  `;

  constructor(props: TextAreaControlProps) {
    super({
      ...props,
      template: TextAreaControl.template
    });
    this.controlElement = this.element.querySelector('textarea') as HTMLTextAreaElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: string) {
    this.state.value = value;
  }

  get value(): string {
    return this.state.value as string;
  }

  private onValue: Listener = (update: ListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = (e.target as HTMLInputElement).value;
    });
  }
}
