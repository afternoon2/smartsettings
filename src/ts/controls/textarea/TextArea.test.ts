import { TextAreaControl } from './TextAreaControl';

let control: TextAreaControl;

const listener = jest.fn();

describe('Text control', () => {
  beforeEach(() => {
    control = new TextAreaControl({
      id: 'id',
      name: 'name',
      options: {
        value: 'test',
        autocomplete: false,
        disabled: false,
      },
      parentElement: document.body,
      listener,
    });
  });

  test('If it matches snapshot after initial load', () => {
    expect(control.element).toMatchSnapshot();
  });

  test('Disable', () => {
    control.disable();
    expect(control.element).toMatchSnapshot();
  });

  test('Enable', () => {
    control.disable();
    control.enable();
    expect(control.controlElement.hasAttribute('disabled')).toBe(false);
    expect(control.element).toMatchSnapshot();
  });

  test('Get and set value', () => {
    control.value = 'Test';
    expect(control.value).toBe('Test');
  });

  test('If it assings user listener correctly', () => {
    expect(listener).toHaveBeenCalledTimes(4);
  });
});