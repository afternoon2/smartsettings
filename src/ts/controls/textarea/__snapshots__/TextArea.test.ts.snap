// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text control Disable 1`] = `
<div
  class="control disabled"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content"
    >
      
    
      <textarea
        class="textarea"
        cols="10"
        rows="4"
        wrap="hard"
      >
              test
    
      </textarea>
      
  
    </div>
  </main>
</div>
`;

exports[`Text control Enable 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content"
    >
      
    
      <textarea
        class="textarea"
        cols="10"
        rows="4"
        wrap="hard"
      >
              test
    
      </textarea>
      
  
    </div>
  </main>
</div>
`;

exports[`Text control If it matches snapshot after initial load 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content"
    >
      
    
      <textarea
        class="textarea"
        cols="10"
        rows="4"
        wrap="hard"
      >
              test
    
      </textarea>
      
  
    </div>
  </main>
</div>
`;
